File: gpt-4-turbo/stage2/serial_construct_reduction_clause.cpp
#include <iostream>
#include <cstdlib> // for rand(), srand()
#include <vector>
#include <numeric> // for std::accumulate
#include <openacc.h>

#ifndef T1
// T1:serial construct reduction clause,V:2.7-3.3
int test1() {
    int err = 0;
    const int N = 1000;
    srand(42); // SEED

    std::vector<int> vec(N);
    for (int i = 0; i < N; ++i) {
        vec[i] = rand() % 100; // Filling vector with random numbers
    }

    int sum_acc = 0;
    // OpenACC operation
    #pragma acc serial copy(vec[0:N]) reduction(+:sum_acc)
    for (int i = 0; i < N; ++i) {
        sum_acc += vec[i];
    }

    // CPU operation for comparison
    int sum_cpu = std::accumulate(vec.begin(), vec.end(), 0);

    // Verifying the results
    if (sum_acc != sum_cpu) {
        std::cerr << "Mismatch: sum_acc = " << sum_acc << ", sum_cpu = " << sum_cpu << std::endl;
        err = 1;
    }

    return err;
}
#endif

int main() {
    int failcode = 0;
    const int NUM_TEST_CALLS = 10; // Example constant
#ifndef T1
    int failed = 0;
    for (int x = 0; x < NUM_TEST_CALLS; ++x) {
        failed = failed + test1();
    }
    if (failed != 0) {
        failcode = failcode + (1 << 0);
    }
#endif
    return failcode;
}

File: gpt-4-turbo/stage2/serial_construct_reduction_clause.f90
!acc_testsuite.Fh
! Mock include file for the purposes of this example.
! Defines NUM_TEST_CALLS.
INTEGER, PARAMETER :: NUM_TEST_CALLS = 1

!T1:serial construct reduction clause,V:2.7-2.3
      LOGICAL FUNCTION test1()
        USE openacc
        IMPLICIT NONE

        INTEGER, PARAMETER :: n = 100
        INTEGER :: i
        INTEGER :: result_acc, result_serial
        INTEGER, DIMENSION(n) :: data

        ! Initialize array.
        DO i = 1, n
           data(i) = i
        END DO

        result_acc = 0
        result_serial = 0

        ! Using OpenACC serial construct with reduction clause.
        !$acc serial reduction(+:result_acc)
        DO i = 1, n
           result_acc = result_acc + data(i)
        END DO

        ! Serial computation for verification.
        DO i = 1, n
           result_serial = result_serial + data(i)
        END DO

        ! Compare the results.
        IF (result_acc == result_serial) THEN
          test1 = .FALSE.
        ELSE
          test1 = .TRUE.
        END IF
      END FUNCTION test1

      PROGRAM main
        IMPLICIT NONE
        INTEGER :: failcode, testrun
        LOGICAL :: failed

        EXTERNAL :: test1

        failed = .FALSE.
        failcode = 0

        DO testrun = 1, NUM_TEST_CALLS
          failed = failed .or. test1()
        END DO
        IF (failed) THEN
          failcode = failcode + 2**0
          failed = .FALSE.
        END IF

        CALL EXIT(failcode)
      END PROGRAM main